### PR TITLE
add a favicon, configured by whitelabel.favicon_url

### LINKED ISSUE #2853: Favicon for CiviForm
It would be great to have the "favicon" for CiviForm match with the city/civic entities' logo for improved site trust and consistency across websites/webpages.

**Describe the solution you'd like**
I would like to have the City of Seattle's logo show on the web browser tab when CiviForm is open. And I imagine other civic entities might want theirs as well!

**Additional context**
The current default favicon is a globe.

**Attached**
Proposed favicon
Current default favicon
Current default Favicon for CiviForm
Favicon for CiviForm

### PR DESCRIPTION
**Pull request by kmcnellis**
### Description

Adds the ability to set a favicon URL (hosted on an external site) to civiform.  By default uses the one at https://civiform.us/favicon.png

## Release notes:

Add the ability to set a favicon to a URL specified by a deployment though the FAVICON_URL env variable.

### Checklist

- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)

Fixes #2853

### CODE REVIEW

File: server/conf/application.conf [hunk 1]
```diff
@@ -52,6 +52,9 @@ whitelabel {
   # The full display name of the civic entity, will use "City of Seattle" if not set.
   civic_entity_full_name = "City of Seattle"
   civic_entity_full_name = ${?WHITELABEL_CIVIC_ENTITY_FULL_NAME}
+
+  favicon_url = "https://civiform.us/favicon.png" # 32x32 pixels
```
Comment (clouser2) [inline server/conf/application.conf:56]: Nit: Consider moving this up to a comment describing the field as well as the fact that the image should be 32x32 pixels as for the above values. If relevant, I'd also note the supported formats (I can't remember what are valid favicon image formats).

File: server/conf/application.test.conf [hunk 2]
```diff
@@ -17,3 +17,5 @@ azure.blob.container = "super cool blob container name"
 azure.blob.account = "my awesome azure account name"
 
 application_status_tracking_enabled = true
+
+auth.applicant_idp = "disabled"
```
Comment (clouser2) [inline server/conf/application.test.conf:21]: Nit: This seems unrelated. Mind splitting it out separately or explicitly updating the PR description to call out this change?

Comment (kmcnellis) [reply server/conf/application.test.conf:21]: It's not related, but it is needed to get the tests to work.  I split it out into https://github.com/seattle-uat/civiform/pull/2927. (I assume this is also what is causing other tests to fail)

File: cloud/deploys/dev_azure/civiform_config.example.sh [hunk 3]
```diff
@@ -49,3 +49,4 @@ export LOGIN_RADIUS_METADATA_URI="https://civiform-staging.hub.loginradius.com/s
 # Name from the Login Radius Dashboard that you set up for local development:
 # https://docs.civiform.us/contributor-guide/developer-guide/dev-azure#setup-login-radius-for-local-development
 export LOGIN_RADIUS_SAML_APP_NAME="CHANGE ME"
```
[the remark this answers is not in the document]
Comment (kmcnellis) [reply cloud/deploys/dev_azure/civiform_config.example.sh:51]: Real, but moved to https://github.com/seattle-uat/civiform/pull/2927 to address it

File: server/conf/application.conf (hunk 1 above)
Comment (kmcnellis) [reply server/conf/application.conf:56]: done

File: server/app/controllers/HomeController.java [hunk 4]
```diff
@@ -86,6 +90,13 @@ public Result playIndex() {
     return ok("public index");
   }
 
+  public Result favicon() {
+    if ((civiformFaviconUrl == null) || civiformFaviconUrl.isBlank()) {
```
[the remark this answers is not in the document]
Comment (kmcnellis) [reply server/app/controllers/HomeController.java:94]: Thanks, I always forget that exists.

File: server/app/views/HtmlBundle.java [hunk 5]
```diff
@@ -163,14 +176,22 @@ private FooterTag renderFooter() {
...
    *   <li>CSS styles
    *   <li>javascript that needs to run immediately
    * </ul>
    */
   private HeadTag renderHead() {
     // TODO: Throw exception if page title is not set.
-    return head().with(title(pageTitle)).with(metadata).with(stylesheets).with(headScripts);
+    return head()
+        .with(title(pageTitle))
+        // The "orElse" value is never used, but it must be used because the
```
[the remark this answers is not in the document]
Comment (kmcnellis) [reply server/app/views/HtmlBundle.java:189]: Yeah, It's sad.  It would be really nice to have a non-evaluating version.

File: server/conf/application.conf [hunk 6]
```diff
@@ -52,6 +52,11 @@ whitelabel {
   # The full display name of the civic entity, will use "City of Seattle" if not set.
   civic_entity_full_name = "City of Seattle"
   civic_entity_full_name = ${?WHITELABEL_CIVIC_ENTITY_FULL_NAME}
+
+  # The URL of a 32x32 or 16x16 pixel [favicon](https://developer.mozilla.org/en-US/docs/Glossary/Favicon)
+  # image, in GIF, PNG, or ICO format.
```
Comment (clouser2) [inline server/conf/application.conf:57]: Looks like we only have route entries for "ico" / "png". Please either remove "gif" from here or add it as a supported route.

**Review by clouser2 — APPROVED** [review]

File: server/app/views/HtmlBundle.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bion) [reply server/app/views/HtmlBundle.java]: i wonder how easy of a PR that would be to the library

File: server/conf/application.conf (hunk 6 above)
Comment (kmcnellis) [reply server/conf/application.conf:57]: That's not really needed - there's 2 things going on here:
1) we set the rel=icon link metadata in the head to the image you specify (the main way to set the favicon)
2) we redirect any browsers who, by default, request favicon from root, to the correct image link. (it's default behaviour, but doesn't follow the spec - https://stackoverflow.com/questions/56222166/prevent-browser-from-trying-to-load-favicon-from-root-directory)

So those routes entries only are needed for the routes that a browser may request, but we can actually return an image in any requested format

Comment (kmcnellis) [reply server/conf/application.conf:57]: Added comments

File: cloud/deploys/dev_azure/civiform_config.example.sh [hunk 7]
```diff
@@ -49,3 +49,4 @@ export LOGIN_RADIUS_METADATA_URI="https://civiform-staging.hub.loginradius.com/s
...
+export FAVICON_URL="https://civiform.us/favicon.png"
```
[the remark this answers is not in the document]
Comment (kmcnellis) [reply cloud/deploys/dev_azure/civiform_config.example.sh:52]: Added to terraform, and made https://github.com/civiform/civiform-deploy/pull/11
